### ISSUE TITLE
Commands in history get mixed up

### ISSUE THREAD

**Issue by wolf99** [Microsoft/WSL#1756]
* A brief description
When traversing back through command history using the up cursor key, shorter commands sometimes show the remains of preceding longer commands.

* Expected results
 When traversing the command history the prompt should show only the text of the command expected, whatever the length.
 For example if I execute the two following commands:
 ~~~shell
 $ rake test:delta
 $ rake gcov:all && rake utils:gcov
 ~~~
 And I traverse back to the `rake test:delta` command, then I expect to see exactly that.

* Actual results (with terminal output if applicable)
 Instead I actually see
~~~shell
$ rake test:delta rake
~~~
 with the cursor placed at the end of `delta` and the trailing ` rake` portion apparently not _actually_ there as far as the prompt is concerned. That is, I can not use the left/right cursor keys to traverse the trailing ` rake`, nor use the Delete key to delete it, nor does it actually execute as part of the traversed to command if I execute that command... it appears to be solely a visual artifact of some kind.

* Your Windows build number
 14393.693

* Steps / All commands required to reproduce the error from a brand new installation 
  - Execute some long-ish commands at the prompt
  - Execute some shorter ones
  - Use the up cursor key to traverse history

* Strace of the failing command
 N/A

* Required packages and commands to install
 N/A

**Comment by aseering** [comment]
I can confirm that this still happens in build 15046.  Though, I don't yet have a specific reproducer.

**Comment by benhillis** [comment]
I have seen this in the past and I believe it's a display issue, not a functionality issue.  What I mean by this is I believe that the console is not clearing the entire screen buffer of the previous command that was being displayed.

@wolf99 - When you get in this state and press enter what gets run?  Is it "rake test:delta" or "rake test:delta rake"?

**Comment by wolf99** [comment]
@benhills, yep, this is a display issue of some sort it only runs the intended `$ rake test:delta`, and not the `$ rake test:delta rake`. As mentioned, it is neither traversable, nor deleteable and for all _functional_ purposes the erroneous text is "not there", it's just *displayed* as being there.

**Comment by benhillis** [comment]
Thanks for confirming - I'm not trying to argue it isn't a bug, "functional" is just the word I used to describe "the text that is displayed versus the command that is actually run."

Hopefully one of our friends from the console team can take a look.

**Comment by zadjii-msft** [comment]
Well. This is definitely close to another known issue that was fixed a month or two back. It definitely was a bug present on 14393.

What troubles me is @aseering's report that this is still happening on 15046. I think that fix was supposed to be in 15014 (according to what's in #1307)

I'll investigate. But repro steps (on a build greater that 15014) would be super helpful. Before it was possible to get into this state by running `apt install`, after scrolling the window down (so that the viewport isn't at 0,0 in the buffer). Unfortunately, that doesn't seem to trigger it anymore.

**Comment by aseering** [comment]
It's very possible that what I am seeing is new or different.  (Though the symptoms seem very similar.)  I will keep trying to come up with a repro.

For what it's worth, I never see it with a new terminal window.  It's always in a window that I have been using for a while, and have run a bunch of commands in.  Also, running `reset` fixes it for me, at least temporarily.  It seems very much like something funny in some state machine somewhere.

**Comment by tucker973** [comment]
I, too, see this issue - unfortunately I wish I knew how to repro, but all I can offer is that I see the same behavior. When using a console window for a "a while," occasionally the bottom lines will mess up. Running `clear` does not seem to address the issue, but exiting and opening a new console does.

Also for what it's worth, I see the problem inside ConEmu as well as simply running Bash by itself.

**Comment by aseering** [comment]
@tucker973 -- I'm curious -- you say that `clear` doesn't work.  Does `reset` work?  (It works for me.  It does more than "clear" does.)  If so, that might (or might not?) be useful for narrowing this down.

**Comment by tucker973** [comment]
I haven't seen the issue since I saw your post about `reset`, so I will give that a shot! I have a few projects to crunch on today, so more than likely I'll see it at least once. I will add to the topic when I have a chance to test. I am on 15046, for the record.

**Comment by wolf99** [comment]
While we're on the topic, and you reference #1762, another issue that I noticed looks to be similar and may as such be related: If I open a file in nano or vim `$ vim foo.c` and then minimise the bash window (from either restore-down size or maximized size) and then return it to restore-down size (or maximize) then only the first line of the file contents are displayed.

The rest of the file is still _functionally_ there, i.e. I can still browse the now invisible contents with the cursor keys and issue commands, but to actually see more than the first line I need to quit vim (not the bash window) and reopen it again. Not sure if it's the same issue but it has the same functional vs display character. Is this evident on 15046 @aseering ? In other editors?

**Comment by aseering** [comment]
@wolf99 -- no, I don't see that on my machine.  Maybe it has been fixed?

**Comment by tucker973** [comment]
I haven't updated beyond 15046 yet, but I just had this happen to me again after running a few things for a while. I was typing a `git commit` message inline, when it started doing the weird behavior. (again, just so everyone is aware, Bash running inside ConEmu, though I've also seen it from running Bash directly) I ran `reset`, did a bunch of `git status` to fill the screen buffer again, and it did still had the same problem. I'll be updating to 15055 this weekend and will check in again to see if this is still happening.

**Comment by zadjii-msft** [comment]
@wolf99 That issue sounds familiar, but it's been a few months since I've done anything like that in the codebase, so I can't recall if I explicitly checked in a fix for that or not. In all likelyhood, it was just fixed by some other change some time since AU came out.

@tucker973 There hasn't been a lot of code delta for conhost over the last few builds, so I doubt that it it's broken in 15046 it'd be fixed in 15055, but it's worth a shot :P

For clarity, I'm 99% certain this is an issue with the scroll margins. That's why `reset` would fix this and not `clear`. `clear` only clears the contents of the buffer. `reset` does a lot more than just clearing the contents of the buffer, including removing the scroll margins.

**Comment by davidmaxwaterman** [comment]
I have been seeing this yonks, and was very keen to try CU/15063.138 in the hope it was fixed in that, but I updated today and it is still there...ie not fixed. I was thinking it was something to do with VT control character support (or lack of) which got quite a big mention in the release notes. I use WSL for all my development, and this is a big PITA...so easy to reproduce, though perhaps not 'on demand'. It seems to happen 'all the time', though perhaps it is only after a while - I'll have to pay attention, and also experiment with the reset command.

**Comment by Alebat** [comment]
**REPRO!!!**

I managed to **reproduce** a version of this problem by the following procedure:
1) open bash.exe (Winkey+R or in cmd)
2) execute *The Script*
3) press ArrowUp 4 times

*The Script*
>  2000  echo 1234567890
 2001  echo 0 9 8 7 6 5 4 3 2 1
 2002  cd
 2003  ls -la
 2004  sudo apt install tomcat7 **OR** sudo apt remove tomcat7

NB: It seems to be the apt loading bar being drawed after `ls -la` of JUST MY home directory (Figure). I think there are some strange sequences of escape sequences...
I tried to reproduce that by an `echo -e` but I didn't manage: `ls -l` **does not work** for me!

![immagine](https://cloud.githubusercontent.com/assets/3705175/25554893/9e167690-2cd9-11e7-83ef-91fb06bc393c.png)

**Comment by zadjii-msft** [comment]
@wolf99 @tucker973 anyone happen to find concrete repro steps?

There's a _really_ good chance this is related to #1984, but I can't be sure without repro steps.

**Comment by davidmaxwaterman** [comment]
Personally, I find this so easy to reproduce that I am staggered that people are asking how.

I can produce it simply by pressing the up arrow key and holding it a while. It doesn't take long for it to happen.

In fact, even if I don't use auto-repeat on the up arrow, I can get it to happen easily. Oddly, it always happens on the same transition. I have these commands in my history :

```
 1982  du -sk * | sort -n
 1983  dd if=/dev/zero of=1GiB.txt count=1024 bs=1048576
 1984  dd if=/dev/zero of=1MiB.txt count=1 bs=1048576
 1985  dd if=/dev/zero of=10MiB.txt count=10 bs=1048576
 1986  dd if=/dev/zero of=100MiB.txt count=100 bs=1048576
 1987  ls -ltr
 1988  cd
...
```

It always happens when I go from 1983 to 1982.

The command line changes from:

```
$ dd if=/dev/zero of=1GiB.txt count=1024 bs=1048576
```

to:

```
$ dd if=/dev/zero ou -sk * | sort -n
```

I doubt that will help you, but I have yet to have a terminal session where I can't easily get this to happen.

(Grasping at straws)...thinking about anything that is odd about my environment...I use vi mode (`set -o vi`) and ` PS1='\e[37;1m\u@\e[35m\W\e[0m\$ '`; also in .bashrc :

```
 11 # don't put duplicate lines or lines starting with space in the history.
 12 # See bash(1) for more options
 13 HISTCONTROL=ignoreboth
 14
 15 # append to the history file, don't overwrite it
 16 shopt -s histappend
 17
 18 # for setting history length see HISTSIZE and HISTFILESIZE in bash(1)
 19 HISTSIZE=1000
 20 HISTFILESIZE=2000
```

**Comment by tucker973** [comment]
Unfortunately, for me I see the issue very intermittently. WSL is an essential part of my workflow, but only for parts of some projects, so I'm not at the terminal often enough to notice a definite pattern. Typically I only notice it when I'm doing a `git commit -m "Long message that should move to a new console line but then starts overwriting itself"`.

**Comment by davidmaxwaterman** [comment]
Oh yeah, when trying to go back to edit some part of a line is hopeless...I've basically given up trying to do that.

**Comment by zadjii-msft** [comment]
@davidmaxwaterman Oh well there it is!

Your PS1 is: `PS1='\e[37;1m\u@\e[35m\W\e[0m\$ '`

What it should be is `PS1='\[\e[37;1m\]\u@\[\e[35m\]\W\[\e[0m\]\$ '`

You need to wrap escape sequences in `\[` and `\]` so that bash knows that they are non-printing characters. Otherwise, bash things the length of your prompt includes those characters. 

See http://tldp.org/HOWTO/Bash-Prompt-HOWTO/bash-prompt-escape-sequences.html:
 > ```
 >               \[     begin a sequence of non-printing characters,
 >                      which could be used to embed a terminal con­
 >                      trol sequence into the prompt
 >               \]     end a sequence of non-printing characters
 > ```

That would explain why it was impossible for me to repro, but happened consistently for you.

@tucker973 I bet you're having the same problem as well.

**Comment by tucker973** [comment]
I'm confused by this, as I haven't modified anything; merely running out-of-the-box WSL. Where would these values be changed? I'm also not seeing it every time I commit something, just occasionally.

**Comment by zadjii-msft** [comment]
@tucker973 Ah, my apologies for not being clearer.

`PS1` is a variable that's used by bash to display the prompt,  the little `zadjii@computer-name$ ` that's at the start of the line.

You can modify the value in your `~/.bashrc` file to change it's appearance or behavior. There are some pretty powerful things you can do with this - Powerline comes to mind as one of the more common ways of augmenting your prompt.

@davidmaxwaterman changed his to just be a little simpler than usual (and magenta for the working directory, that's the `\e[35m\W\e[0m` bit). That's what was causing this for him. Not sure why this would affect you though if you haven't modified your `.bashrc`.

You say you see it occasionally while writing commit messages with `git commit -m <message>`, correct? I'll play with that, see if I can get a repro.

**Comment by tucker973** [comment]
Ah, the only thing I've done with .bashrc is look at it and realize the aliases were stored in their own file, so it should be exactly how it shipped. 

What's odd is it doesn't seem to matter how much work I've done - for example, today I've had the same prompt open making commits all day and haven't seen it happen once. Other days, I'll hop on to do just a few things and it happens.

It's more of a minor annoyance than anything, but I will continue to try to keep an eye on it to see if I can provide any additional insight! Thanks to you and the team for putting out such a killer product - if this is the worst bug I encounter, I'm happy!

**Comment by davidmaxwaterman** [comment]
@zadjii-msft Well, that's odd, isn't it? I wonder where I got the PS1 string from - I guess . Doing some searches, it does seem pretty common to *not* have them. Is the WSL terminal particularly sensitive or something, I wonder? Also, it seems strange that it doesn't happen *every* time I move back/forward in history.

Anyway, it does seem effective and so we have a solution - great! :)

Personally, I consider this fixed (but perhaps this isn't the only issue).

Now, if only we could have the fix for ctrl-z :(

**Comment by zadjii-msft** [comment]
@davidmaxwaterman Could you open a new issue describing what's wrong with ctrl+z? I don't believe there are any existing issues with it so I'd love to know if there is one.

(and also not to dilute an existing thread with a tangent :) )

@tucker973 That is odd. keep an eye on it. Maybe it'll go away on it's own :P Happy to make something you enjoy 👍

**Comment by benhillis** [comment]
@zadjii-msft - Ctrl+Z issue #1569

**Comment by zadjii-msft** [comment]
Oh goodness, my bad. I must be on an older build than insiders :P

**Comment by asampal** [comment]
I hit this bug after resizing the console vertically. That also messed up output from "apt-get install" which I thought wasn't that big a deal since I had done it while output was in progress. However the way moving through history got borked, as described in this issue, was worse. Had to restart the console.

I'm on CU with bash 4.3.48(1)

**Comment by Alebat** [comment]
## SSH sessions
In SSH sessions it always happen to me what explained in #1307:
### Example:
If I arrow up I get for instance the following command:
`make install|` with "|" indicating the cursor.
If I stroke five left arrows:
`make in|stall`
If I type "who":
`make inwho|ll`
If I press enter the executed command is:
`make inwhostall`

Just tried now towards a `FreeBSD 11.1-RELEASE-p1 (GENERIC)`

**Comment by zadjii-msft** [comment]
@Alebat what Windows build are you on?

Also, can you post the output of the following two commands?

`infocmp`

`echo $PS1`

**Comment by amasover** [comment]
@Alebat @zadjii-msft I also see this behavior when ssh'ing from Ubuntu/WSL on Windows build 1803 build 17134.48 to a server running FreeBSD 11.1-STABLE (a FreeNAS machine).

**Comment by amasover** [comment]
Apologies for my barrage of comments.

I just noticed that, on the FreeBSD side, my shell was set to `csh`. When I change my shell preference to `bash` (through account settings in the FreeNAS UI), I don't see this behavior anymore.

**Comment by bitcrazed** [comment]
Thanks for the discussion.

Closing this issue since:
1. This issue has likely been fixed in recent Win10 builds
1. This is the WSL issues repo, but this is an issue in Console which has its own [Console GitHub Repo](https://github.com/Microsoft/console)
1. GitHub doesn't allow issues to be moved between repos, preserving posters' identity :(

If you have further asks/issues, please file new issues on our [Console GitHub Repo](https://github.com/Microsoft/console).
